<!DOCTYPE html>
<html lang="en"><head>
<link href="../../logo.png" rel="icon" type="image/png">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-07ba0ad10f5680c660e360ac31d2f3b6.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.33">

  <meta name="author" content="Marie">
  <meta name="author" content="Kayleigh">
  <meta name="author" content="Xuyang">
  <meta name="dcterms.date" content="2024-05-22">
  <title>Robot Genome ‚Äì PGDIS 2024</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-15ee1b0a1723982d7b83a3601d199550.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="venue.JPG" data-background-opacity="0.5" class="quarto-title-block center">
  <h1 class="title">PGDIS 2024</h1>
  <p class="subtitle">üìÖ May 6-8 2024 üá≤üáæ Kuala Lumpur, Malaysia</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Marie 
</div>
</div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Kayleigh 
</div>
</div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Xuyang 
</div>
</div>
</div>

  <p class="date">2024-05-22</p>
</section>
<section id="conference-agenda" class="slide level2" data-background-video="slidebg.m4v" data-background-video-loop="true" data-background-video-muted="true" data-background-video-size="cover" data-background-opacity="0.7">
<h2>Conference Agenda</h2>
<div class="highlight-last">
<ul>
<li class="fragment">Day 0 (Sun): Pre-conference workshop (Thermo Fisher PGT solutions, carrier screening, methylation, TGS)</li>
<li class="fragment">Day 1 (Mon): Advances in PGT + Contraversies in PGT</li>
<li class="fragment">Day 2 (Tue): Contraversies in PGT + Embryo selection + niPGT + genetic counselling + Gala dinner</li>
<li class="fragment">Day 3 (Wed): Quality in IVF + Inovations in Preimplantation Genetics</li>
</ul>
</div>
</section>
<section id="thermo-fisher-scientifics-ngs-solutions-for-pgt" class="slide level2">
<h2><u>Thermo Fisher Scientific</u>‚Äôs NGS Solutions For PGT</h2>
<div style="font-size: 0.8em;">
<p>Whole Workflow: Lab Prep + Sequencing + Analysis + Reporting</p>
<ul>
<li>Library Prep <span style="color: gray">(DNA extraction, WGA and barcoding)</span> and Template Prep <span style="color: gray">(automated clonal amplification)</span>
<ul>
<li><a href="https://www.thermofisher.com/order/catalog/product/4484177" class="external">Ion Chef‚Ñ¢ System</a></li>
</ul></li>
<li>Sequencing (Ion Semiconductor Sequencing):
<ul>
<li><a href="https://www.thermofisher.com/no/en/home/life-science/sequencing/next-generation-sequencing/ion-torrent-next-generation-sequencing-workflow/ion-torrent-next-generation-sequencing-run-sequence/ion-s5-ngs-targeted-sequencing.html" class="external">Ion GeneStudio‚Ñ¢ S5 System</a></li>
<li><a href="https://www.thermofisher.com/no/en/home/life-science/sequencing/next-generation-sequencing/ion-torrent-next-generation-sequencing-workflow/ion-torrent-next-generation-sequencing-run-sequence/ion-torrent-genexus-system.html" class="external">Ion Torrent Genexus‚Ñ¢ System</a></li>
</ul></li>
<li>Analysis and Reporting:
<ul>
<li><a href="https://ionreporter.thermofisher.com/ir/" class="external">Ion Reporter‚Ñ¢ Software</a></li>
</ul></li>
</ul>
</div>
</section>
<section id="pgt-a" class="slide level2 smaller">
<h2>PGT-A</h2>
<h3 id="ion-reproseq-pgs"><a href="https://www.thermofisher.com/no/en/home/life-science/sequencing/next-generation-sequencing/ion-torrent-next-generation-sequencing-applications/reproductive-health-research/preimplantation-genetic-testing/preimplantation-genetic-screening.html">Ion ReproSeq‚Ñ¢ PGS</a></h3>
<ul>
<li>Detect whole-chromosome, mosaic, and small copy number events with just 6 pg of DNA from single or multiple cells</li>
<li>Multiple kit configurations to analyze from 16 to 192 samples in a single run</li>
<li>Mosaicism detection, gender masking, and improved data plotting (for easier data interpretation)</li>
<li>Together with the <em>Ion AmpliSeq‚Ñ¢ Polyploidy Panel</em> can confidently identify triploidy 69, XXX, assess potential maternal contamination, and track sibling samples, preventing sample mix-ups.</li>
<li>Turnaround time for the Ion ReproSeq PGS Kit 16 samples/run is &lt;10 hours with &lt;2 hours of hands on time.</li>
</ul>
</section>
<section id="pgt-m" class="slide level2">
<h2>PGT-M</h2>
<h4 id="pgd-seq-panel-and-reagent-kits"><a href="https://www.thermofisher.com/no/en/home/life-science/sequencing/next-generation-sequencing/ion-torrent-next-generation-sequencing-applications/reproductive-health-research/preimplantation-genetic-testing/targeted-ngs-monogenic.html">PGD-SEQ‚Ñ¢ Panel and Reagent Kits</a></h4>
<ul>
<li>&gt;500 gene panels for common and rare monogenic diseases for IVF</li>
<li>Complete kit reagents and targeted PGT-M library to analyze 15 samples</li>
<li>Comprehensive familial carrier status using proprietary linkage analysis software</li>
<li><blockquote>
<p><strong>Simultaneous PGT-M and PGT-A using a single embryo biopsy</strong></p>
</blockquote></li>
</ul>
</section>
<section id="epigenetics-testing" class="slide level2">
<h2>Epigenetics Testing</h2>
<p>Preimplantation DNA methylation Screening (<strong>PIMS</strong>)</p>
<ul>
<li>Current PGT mainly focus on genetic causes</li>
<li>Only about 1/3 of birth defects can find reasons</li>
<li>Both genetic and epigenetic factors can cause birth defects</li>
<li>Compared to standard PGT-A, PIMS increases pregnancy rate and live birth rate by 10-15%</li>
<li>Exclusive methylation-based <strong>artificial intelligence</strong> to improve accuracy and to avoid subjectivity</li>
</ul>
</section>
<section id="mosaisism" class="slide level2 smaller">
<h2>Mosaisism</h2>
<ul>
<li>IRMET (international registry of mosaic embryo transfer) (irmet.net)</li>
<li>Kilder til mossaikk: endogene, indusert, artifaktisk</li>
<li>20-30% mosaikk-tilstander er alt for h√∏yt - m√• da finne avvik (RMA, lab/operat√∏r, osv)</li>
<li>Gold standard: celle linjer (blinded?), brukes for √• validere og definere ‚Äúcut off‚Äù/st√∏y hos oss</li>
<li>Kj√∏re OM alle segmentale aneueploidier (Rana et. Al, 2023; D. Marin)</li>
<li>Hvordan rapporteres dette?</li>
<li>Det er caser med uheldige utfall, med disse er undertrykt/ikke publisert pga juridiske problemer</li>
</ul>
</section>
<section id="mosaikk" class="slide level2">
<h2>Mosaikk</h2>

<img data-src="mosaikk.png" class="r-stretch"></section>
<section id="mosaikk-fort." class="slide level2">
<h2>Mosaikk fort.</h2>

<img data-src="mosaikk_fort.png" class="r-stretch"></section>
<section id="mosaikk-fort.-1" class="slide level2">
<h2>Mosaikk fort.</h2>

<img data-src="mosaikk_fort_1.png" class="r-stretch"></section>
<section id="mosaikk-rapportering" class="slide level2">
<h2>Mosaikk, rapportering</h2>
<p><span style="font-size: 0.5em;">Mosaikk-embryo klassifisering for et redusert embryotap, v/Steve Grkovic (genea, Australia)</span> <img data-src="mosaikk_rapport.png"></p>
</section>
<section id="mosaikk-rapportering-1" class="slide level2">
<h2>mosaikk, rapportering</h2>

<img data-src="mosaikk_rapport_1.png" class="r-stretch"></section>
<section id="mosaikk-rapportering-2" class="slide level2">
<h2>mosaikk, rapportering</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="mosaikk_rapport_2.png"></p>
</div><div class="column" style="width:50%;">
<p><img data-src="mosaikk_rapport_3.png"></p>
</div></div>
</section>
<section id="medlem-av-pgdis-david-cram-om-mosaikk" class="slide level2 smaller">
<h2>Medlem av PGDIS (David Cram) om mosaikk</h2>
<ul>
<li>Hvordan definere cut-off for ‚Äúlav-‚Äù og ‚Äúh√∏y‚Äù-gradig mosaikk: &nbsp;
<ul>
<li>Alt. 1: bruke 50% som cut off&nbsp;</li>
<li>Alt. 2 er √• ta st√∏yniv√• satt av laboratoriet til betrakning. &nbsp;</li>
</ul></li>
<li>Hvis st√∏ygrense er 20% for √• definere euploid vil:&nbsp;
<ul>
<li>M√•ling av 30% kan v√¶re s√• h√∏y som 50% eller lav som 10%&nbsp;</li>
<li>M√•ling av 40% kan v√¶re s√• h√∏y som 60% eller lav som 20%&nbsp;</li>
<li>M√•ling av 50% kan v√¶re s√• h√∏y som 70% eller lav som 30%&nbsp;</li>
</ul></li>
</ul>

<img data-src="transfer_priority.jpg" class="r-stretch"></section>
<section id="mosasic-transfer" class="slide level2">
<h2>Mosasic transfer</h2>
<ul>
<li>&gt;99% of live births from mosaic embryos transfer have no detectable abnormalities (Note the prerequisites of ongoing pregnancy, i.e.&nbsp;no implantation failure, no miscarriage)</li>
<li>Low-level mosaicism has a better outcome for implantation and pregnancy. But mosiac level ‚Äúcurrenly‚Äù does not predict the risk of being in the &lt;1% of live births with abnormalities.</li>
</ul>
</section>
<section id="oversikt-platformer" class="slide level2">
<h2>Oversikt platformer</h2>
<ul>
<li>PGT-A: trenger kun kopitall, vanligvis med lav-dekning sekvensering (&lt;0.1X&nbsp;dekning), ca 10Mb st√∏rrelse&nbsp;&nbsp;</li>
<li>PGT-SR: lav-dekning sekvensering kan detektere ubalanser. <strong>Trenger SNP-analyse</strong> for √• skille b√¶rer fra normale&nbsp;</li>
<li>PGT-M: linkage eller <strong>SNP-analyse</strong> (haplotyping eller direkte deteksjon av mutasjonen)&nbsp;</li>
<li>PGT-P: trenger <strong>SNP-analyse</strong></li>
<li>PGT-WGS: alt over, trenger <strong>SNP-analyse</strong></li>
</ul>
</section>
<section id="pgt-metoder" class="slide level2">
<h2>PGT metoder</h2>
<h6 id="snp-array-genotyping">SNP array Genotyping</h6>
<p>Karyomapping, Haplarithmisis. High throughput SNP array, HaploSew, APCAD, Genome pridication PGT-PS, Hapltype-Aware, Whole genome prediction</p>
<h6 id="ngs-genotyping">NGS genotyping</h6>
<p>PGT Complete, One PGT, Genotyping by sequencing, Cheb et al., HaploPGT, S-HaploSeek</p>
<h6 id="sekvensering-og-target-snps">Sekvensering og target SNPs</h6>
<p>PGT-Seq (Revvity), OneGene PGT</p>
</section>
<section id="utf√∏rer-wgs-embryo-screening" class="slide level2 smaller">
<h2>Utf√∏rer WGS embryo screening</h2>
<ul>
<li>Gattaca Genomics (Florida, USA)</li>
<li>Orchid: genpanel med 1,400 gener for arvet patogene sykdommer og polygen risiko)</li>
<li>Panacea ‚Äì GenomeScreen TM: &gt;2,500 sykdomsfremkallende gener for arvelige eller nye de novo-mutasjoner (GenEmbryomics/Progensis Inc)
<ul>
<li>Foredlere ved x30 dybde og embryoer ved 50x dybde</li>
<li>Omfattende variant filtrering for √• eliminere FP</li>
<li>Variant annotering utf√∏rt fra &gt;50 annoteringskilder, algoritmer for prediksjon av patogenisitet, og ACMG guidelines</li>
<li>Benchmarking: Genome in a Bottle, flow-sorted GIAB celle linje, sammenlignet gDNA og amplifisert DNA (RepliG), vurdert med hap.py/VCFEval (Conrad Nat. Genet. (2011))</li>
<li>Binning approach (hadde problemer med targeted CNV)</li>
<li>De nova mutasjoner: n√∏yaktighet beregnet etter filtrering men f√∏r annotatering</li>
<li>niPGT: lav og variable DNA fragmenter ‚Äì&gt; ikke s√• robust</li>
</ul></li>
</ul>
</section>
<section id="quality-in-ivf" class="slide level2">
<h2>Quality in IVF</h2>
<ul>
<li>Juno Genetics (v/Dagan Wells) tar video av overf√∏ringer.</li>
<li>Cytogenomix laboratory in Malaysia: Utf√∏rer shipping validation for IVF-lab som har forh√∏yet amplifiseringsfeil.&nbsp;</li>
<li>For NGS-basert metode:&nbsp;amplifisere mtDNA fra WGA-produkt (mtDetect Web App, skybasert m/server i USA).</li>
</ul>
</section>
<section id="section" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_081023.jpg" class="r-stretch"></section>
<section id="section-1" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_081206.jpg" class="r-stretch"></section>
<section id="section-2" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_081312.jpg" class="r-stretch"></section>
<section id="section-3" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_081511.jpg" class="r-stretch"></section>
<section id="section-4" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_081648.jpg" class="r-stretch"></section>
<section id="section-5" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_081754.jpg" class="r-stretch"></section>
<section id="section-6" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_082146.jpg" class="r-stretch"></section>
<section id="section-7" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_082337.jpg" class="r-stretch"></section>
<section id="section-8" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_082607.jpg" class="r-stretch"></section>
<section id="section-9" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_082734.jpg" class="r-stretch"></section>
<section id="section-10" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_083011.jpg" class="r-stretch"></section>
<section id="de-novo-mutations-or-incomplete-pedigree" class="slide level2">
<h2>de novo mutations or incomplete pedigree</h2>
<ul>
<li>Nesten ingen som gj√∏r dette per i dag, var diskusjoner om hvorvidt man skal begynne √• rapportere dette. Joris fra Leuven snakket om dette. Hvis vi begynner √• rapportere dette for embryo vil det v√¶re et paradigmeskifte. De Novo krever s√• mange som mulig embryobiopsier for √• skille ulike haplotyper.</li>
<li>Hva ender man til slutt med √• kunne sette inn ved WES/WGS for DeNovo mutasjoner?</li>
<li>Long-read sequencing (and direct haplotyping) when incomplete pedigree, de novo mutations (Yikon).</li>
</ul>
</section>
<section id="ai" class="slide level2">
<h2>AI</h2>
<h4 id="re-classification-of-mosaic-embryos-coopersurgical">re-classification of mosaic embryos (CooperSurgical)</h4>
<ul>
<li>Mosasic Resolution by Decompositon</li>
<li>Clinical and theoretical decomposition to estimate mosaicism</li>
<li>Copy number variation (CNV) and allelic balance</li>
</ul>
<h4 id="gradingselection-embryos">Grading/Selection Embryos</h4>
</section>
<section id="section-11" class="slide level2">
<h2></h2>

<img data-src="IMG_20240507_121808.jpg" class="r-stretch"></section>
<section id="section-12" class="slide level2">
<h2></h2>

<img data-src="IMG_20240507_122845.jpg" class="r-stretch"></section>
<section id="non-invasive-pgt-nipgt" class="slide level2">
<h2>Non-invasive PGT (niPGT)</h2>
<ul>
<li>During in vitro development, mostly from day 4 to 6, embryo cell-free DNA (cfDNA) is released to the culture medium, with higher concentrations as the number of cells increase at blastocyst.</li>
<li>informativity rates on D6: 98.8% (D5: 81.8%)</li>
<li>ploidy concordance TE on D6: 89.1% (D5: 76%)</li>
<li>concordance rate with full frozen blastocyst/ICM: up to 93.7% D6/7</li>
</ul>
</section>
<section id="pgt-a-1" class="slide level2">
<h2>PGT-A</h2>
<p>PGTai 2.0 (<span style="color:blue;">CooperSurgical</span>) VS PGTseq-A (<span style="color:blue;">JunoGenetis</span>) VS PG-Seq Rapid v2 Kit (<span style="color:blue;">Revvit</span>)</p>
<p>PGTseq-A: <span style="font-size: 0.4em;">NGS of thousands of sites on each chromosome to get accurate copy number information and SNPs</span></p>
<div class="r-stack">
<p><img data-src="copy_number.png" class="fragment"></p>
<p><img data-src="snps.png" class="fragment"></p>
</div>
</section>
<section id="pgt-a-2" class="slide level2">
<h2>PGT-A</h2>
<p>PGTai 2.0 (CooperSurgical) VS PGTseq-A (JunoGenetis)</p>
<p>PGTai 2.0: <span style="font-size: 0.5em;">Also NGS and combines copy number information with SNPs, but powered by AI</span> <img data-src="PGTai2.0.png"></p>
</section>
<section id="section-13" class="slide level2">
<h2></h2>

<img data-src="IMG_20240506_164838.jpg" class="r-stretch"></section>
<section id="section-14" class="slide level2">
<h2></h2>

<img data-src="IMG_20240506_165126.jpg" class="r-stretch"></section>
<section id="section-15" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_083335.jpg" class="r-stretch"></section>
<section id="section-16" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_083812.jpg" class="r-stretch"></section>
<section id="section-17" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_084053.jpg" class="r-stretch"></section>
<section id="section-18" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_084201.jpg" class="r-stretch"></section>
<section id="section-19" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_084333.jpg" class="r-stretch"></section>
<section id="section-20" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_084537.jpg" class="r-stretch"></section>
<section id="section-21" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_084634.jpg" class="r-stretch"></section>
<section id="section-22" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_084708.jpg" class="r-stretch"></section>
<section id="section-23" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_102059.jpg" class="r-stretch"></section>
<section id="section-24" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_121754.jpg" class="r-stretch"></section>
<section id="section-25" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_122130.jpg" class="r-stretch"></section>
<section id="section-26" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_122241.jpg" class="r-stretch"></section>
<section id="section-27" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_122331.jpg" class="r-stretch"></section>
<section id="section-28" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_122523.jpg" class="r-stretch"></section>
<section id="section-29" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_122633.jpg" class="r-stretch"></section>
<section id="section-30" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_122838.jpg" class="r-stretch"></section>
<section id="section-31" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_123015.jpg" class="r-stretch"></section>
<section id="section-32" class="slide level2">
<h2></h2>

<img data-src="IMG_20240508_123122.jpg" class="r-stretch"></section>
<section id="kontaktpersoner" class="slide level2">
<h2>Kontaktpersoner</h2>
<ul>
<li><em>Carmen</em> fra iGenomix Valencia</li>
<li><em>Edith Coonen</em> (fikk info om at hun var go-to for PGT-SR)</li>
<li><em>David Kubicek</em> Poster 29</li>
<li><em>Aaron Chen</em> fra Malaysia som har brukt Karyomapping og Embryomap i mange √•r</li>
<li><em>Silja Lu</em> fra Yikon (lusijia@yikongenoics.com)</li>
</ul>
</section>
<section id="section-33" class="slide level2">
<h2></h2>
<div class="r-stack">
<img data-src="threeofus.jpg" class="fragment r-stretch">
<img data-src="discussion.jpeg" class="fragment r-stretch">
<img data-src="ask.jpg" class="fragment r-stretch">
<img data-src="lunch.jpg" class="fragment r-stretch">
<img data-src="queen_come.jpg" class="fragment r-stretch">
<img data-src="queen1.jpg" class="fragment r-stretch">
<img data-src="dance.jpg" class="fragment r-stretch">
</div>


</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': false,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true,"theme":"chalkboard","boardmarkerWidth":5,"chalkWidth":2,"chalkEffect":1},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: true,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/robotgenome\.com");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
              // target, if specified
              link.setAttribute("target", "_blank");
              if (link.getAttribute("rel") === null) {
                link.setAttribute("rel", "noopener");
              }
              // default icon
              link.classList.add("external");
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>